### PR TITLE
Use Amazon Linux 2 as base AMI

### PR DESCRIPTION
**Pull request by lox**
This uses the new [Amazon Linux 2](https://aws.amazon.com/amazon-linux-2/release-notes/) as a base for the agent machines.

This will allow us to use systemd!

### CODE REVIEW

**Comment by tduffield** [comment]
This would be very VERY nice to have! Let me know if there's anything I can do to help. We're currently maintaining our own fork of your packer stuff because we needed systemd, so we're happy to give back what learnings we can.

**Comment by lox** [comment]
Learnings would be really helpful @tduffield! I've already run into a few bugs where the agent doesn't have environment that it was expecting, namely `PWD` and `TERM`.

Anything else you've encountered that would be helpful?

**Comment by lox** [comment]
@sj26 / @toolmantim feels on merging this and releasing it as Elastic Stack v4?

**Comment by toolmantim** [comment]
I don't have any feels, but I trust you and @sj26!

**Comment by sj26** [comment]
I have so many feels about this, and they're all _excited_ :clap:

I'll try spinning up a stack next week to give it a go.

File: packer/conf/bin/bk-install-elastic-stack.sh [hunk 1]
```diff
@@ -129,9 +129,8 @@ LIFECYCLED_SNS_TOPIC=${BUILDKITE_LIFECYCLE_TOPIC}
 LIFECYCLED_HANDLER=/usr/local/bin/stop-agent-gracefully
 EOF
 
-# my kingdom for a decent init system
-start lifecycled || true
-service awslogs restart || true
+systemctl start lifecycled
+systemctl start awslogsd
```
Comment (sj26) [inline packer/conf/bin/bk-install-elastic-stack.sh]: :clap:

File: packer/conf/bin/bk-install-elastic-stack.sh [hunk 2]
```diff
@@ -145,9 +144,8 @@ if ! docker ps ; then
 fi
 
 for i in $(seq 1 "${BUILDKITE_AGENTS_PER_INSTANCE}"); do
-	cp /etc/buildkite-agent/init.d.tmpl "/etc/init.d/buildkite-agent-${i}"
-	service "buildkite-agent-${i}" start
-	chkconfig --add "buildkite-agent-${i}"
+  systemctl enable "buildkite-agent@${i}"
+  systemctl start "buildkite-agent@${i}"
```
Comment (sj26) [inline packer/conf/bin/bk-install-elastic-stack.sh]: :clap::clap::clap:

File: packer/conf/buildkite-agent/systemd/buildkite-agent@.service [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+Description=Buildkite Agent (%i)
+Documentation=https://buildkite.com/agent
+After=syslog.target
+After=network.target
+After=docker.service
+
+[Service]
+Type=simple
+User=buildkite-agent
+Environment="HOME=/var/lib/buildkite-agent"
```
Comment (sj26) [inline packer/conf/buildkite-agent/systemd/buildkite-agent@.service]: You might like to add `USER=buildkite-agent`, too. Maybe there are others?

File: packer/conf/docker/daemon.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+  "userns-remap": "buildkite-agent",
```
Comment (sj26) [inline packer/conf/docker/daemon.json]: Is this file meant to be the not remap version?

File: packer/scripts/install-awslogs.sh [hunk 5]
```diff
@@ -7,10 +7,8 @@ sudo yum update -y -q
...
 echo "Adding awslogs config..."
-
 sudo mkdir -p /var/awslogs/state
 sudo cp /tmp/conf/awslogs/awslogs.conf /etc/awslogs/awslogs.conf
 
-echo "Configure it to run on startup..."
-
-sudo chkconfig awslogs on
+echo "Configure awslogsd to run on startup..."
+sudo systemctl enable awslogsd.service
```
Comment (sj26) [inline packer/scripts/install-awslogs.sh]: They don't build this in yet? 🤔

File: packer/conf/bin/bk-install-elastic-stack.sh (hunk 2 above)
Comment (sj26) [reply packer/conf/bin/bk-install-elastic-stack.sh]: (This means `systemctl stop "buildkite-agent@*"` should work!)

File: packer/conf/buildkite-agent/systemd/buildkite-agent@.service (hunk 3 above)
Comment (lox) [reply packer/conf/buildkite-agent/systemd/buildkite-agent@.service]: Done!

File: packer/scripts/install-awslogs.sh (hunk 5 above)
Comment (lox) [reply packer/scripts/install-awslogs.sh]: Nope, not by default.

Comment (sj26) [reply packer/scripts/install-awslogs.sh]: 😢

**Comment by sj26** [comment]
I'm so excited.

How can I easily spin up a copy of this, @lox?

**Comment by lox** [comment]
I mostly rely on CI for testing these @sj26, could you create an Cloudformation stack with https://s3.amazonaws.com/buildkite-aws-stack/amazon-linux-2/aws-stack.yml though!

**Comment by lox** [comment]
@toolmantim and @sj26, I'm thinking of merging this into master and testing it out for a while before rolling a v4. Any objections?

**Comment by toolmantim** [comment]
🚀

**Comment by sj26** [comment]
@lox do it

**Pull request merged**